File: 2025-01-08/Majority_Element_II.java
// 229. Majority Element II

// https://leetcode.com/problems/majority-element-ii/description/

import java.util.ArrayList;
import java.util.List;

public class Majority_Element_II {
    public static List<Integer> majorityElement(int[] nums) {
        List<Integer> ans = new ArrayList<>();
        int fst = 0, snd = 0, c1 = 0, c2 = 0;

        for(int num : nums) {
            if(num == fst)
                c1++;
            else if(num == snd)
                c2++;
            else if(c1 == 0) {
                fst = num;
                c1++;
            }
            else if(c2 == 0) {
                snd = num;
                c2++;
            }
            else {
                c1--;
                c2--;
            }
        }

        c1 = 0;
        c2 = 0;

        for(int num : nums) {
            if(num == fst)
                c1++;
            else if(num == snd)
                c2++;
        }

        if(c1 > nums.length / 3)
            ans.add(fst);
        if(c2 > nums.length / 3)
            ans.add(snd);

        return ans;
    }
    public static void main(String[] args) {
        int[] arr = new int[]{4,2,1,1};
        System.out.println(majorityElement(arr));
    }
}

File: 2026-01-23/Majority_Element_II.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        if not nums:
            return []

        cand1 = cand2 = None
        c1 = c2 = 0
        for num in nums:
            if num == cand1:
                c1 += 1
            elif num == cand2:
                c2 += 1
            elif c1 == 0:
                cand1, c1 = num, 1
            elif c2 == 0:
                cand2, c2 = num, 1
            else:
                c1, c2 = c1 - 1, c2 - 1

        ans = []
        for c in [cand1, cand2]:
            if c is not None and nums.count(c) > len(nums)/3:
                ans.append(c)

        return ans


sol = Solution()
# print(sol.majorityElement([3,3,3]))
print(sol.majorityElement([1,2,1,2,2,3,3,1]))
print(sol.majorityElement([1,2,3]))